File: frontend/src/services/ocr.ts
/**
 * OCR Service - API client for PDF-to-Test import pipeline.
 *
 * Provides:
 * - PDF upload and job creation
 * - Job status and progress tracking
 * - Extracted questions management
 * - Review workflow
 * - Import to test module
 */

import api from '@/lib/axios';

// ===== Types =====

export type OCRJobStatus =
  | 'pending'
  | 'uploading'
  | 'processing'
  | 'structuring'
  | 'review'
  | 'importing'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type OCRProvider = 'deepinfra' | 'openai' | 'hybrid' | 'replicate' | 'openrouter';

/** OCR quality setting - only affects OpenRouter vision model */
export type OCRQuality = 'fast' | 'quality';

export type QuestionReviewStatus =
  | 'pending'
  | 'approved'
  | 'rejected'
  | 'needs_edit'
  | 'imported';

export type QuestionType =
  | 'multiple_choice'
  | 'multiple_choice_math'
  | 'student_produced_response';

export type QuestionDomain =
  | 'algebra'
  | 'advanced_math'
  | 'geometry_trigonometry'
  | 'problem_solving_data_analysis'
  | 'craft_and_structure'
  | 'information_and_ideas'
  | 'expression_of_ideas'
  | 'standard_english_conventions';

export type QuestionDifficulty = 'easy' | 'medium' | 'hard';

export interface OCRJob {
  id: number;
  status: OCRJobStatus;
  pdf_filename: string;
  total_pages: number;
  processed_pages: number;
  question_pages: number;
  skipped_pages: number;
  extracted_questions: number;
  approved_questions: number;
  imported_questions: number;
  progress_percent: number;
  ocr_provider: OCRProvider;
  estimated_cost_cents: number;
  actual_cost_cents: number;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  created_at: string;
  target_module_id: number | null;
}

export interface OCRJobListResponse {
  items: OCRJob[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface QuestionOption {
  id: string;
  text: string;
  image_url?: string | null;
}

export interface TableData {
  headers: string[];
  rows: string[][];
  title?: string | null;
}

export interface ExtractedQuestion {
  id: number;
  job_id: number;
  source_page_number: number;
  review_status: QuestionReviewStatus;
  extraction_confidence: number;
  answer_confidence: number;
  question_text: string;
  question_type: QuestionType;
  passage_text: string | null;
  chart_title: string | null;
  chart_data: string | null;
  table_data: TableData | null;  // New structured table format
  options: QuestionOption[] | null;
  correct_answer: string[] | null;
  needs_answer: boolean;
  explanation: string | null;
  difficulty: QuestionDifficulty | null;
  domain: QuestionDomain | null;
  needs_image: boolean;
  question_image_url: string | null;
  question_image_s3_key: string | null;
  image_extraction_status: string | null;
  validation_errors: string[] | null;
  created_at: string;
}

export interface ExtractedQuestionListResponse {
  items: ExtractedQuestion[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface QuestionUpdateData {
  question_text?: string;
  question_type?: QuestionType;
  passage_text?: string;
  options?: QuestionOption[];
  correct_answer?: string[];
  explanation?: string;
  difficulty?: QuestionDifficulty;
  domain?: QuestionDomain;
  review_status?: QuestionReviewStatus;
}

export interface ImportResult {
  imported: number;
  errors: string[];
  test_id?: number;
}

// ===== Test Configuration Types =====

export interface ModuleConfig {
  section: 'reading_writing' | 'math';
  module: 'module_1' | 'module_2';
  difficulty: 'standard' | 'easier' | 'harder';
  question_start: number;
  question_end: number;
  time_limit_minutes: number;
}

export interface TestConfig {
  test_title: string;
  test_type: 'full_test' | 'section_test' | 'module_test';
  section: 'reading_writing' | 'math' | null;
  modules: ModuleConfig[];
  is_published: boolean;
  is_premium: boolean;
}

export interface TestConfigResponse {
  job_id: number;
  test_config: TestConfig;
  estimated_questions: number;
}

export interface ImportWithTestRequest {
  test_config: TestConfig;
  question_ids?: number[];
}

export interface ImportWithTestResponse {
  test_id: number;
  test_title: string;
  modules_created: number;
  questions_imported: number;
  errors: string[];
}

export interface JobProgressEvent {
  type: 'progress' | 'complete' | 'error';
  data: {
    processed_pages?: number;
    total_pages?: number;
    percent?: number;
    question_pages?: number;
    skipped_pages?: number;
    extracted_questions?: number;
    status?: OCRJobStatus;
    error_message?: string;
    message?: string;
  };
}

// ===== Passage Types =====

export interface ExtractedPassage {
  id: number;
  job_id: number;
  source_page_number: number | null;
  title: string | null;
  content: string;
  source: string | null;
  author: string | null;
  word_count: number | null;
  figures: Array<{ s3_key?: string; url?: string; alt?: string; caption?: string }> | null;
  genre: string | null;
  topic_tags: string[] | null;
  review_status: QuestionReviewStatus;
  extraction_confidence: number;
  linked_questions_count: number;
  created_at: string;
}

export interface ExtractedPassageListResponse {
  items: ExtractedPassage[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface PassageUpdateData {
  title?: string;
  content?: string;
  source?: string;
  author?: string;
  genre?: string;
  topic_tags?: string[];
  review_status?: QuestionReviewStatus;
}

// ===== Phase 6: Error Handling Types =====

export interface FailedPageInfo {
  page_number: number;
  error_message: string | null;
  retry_count: number;
  last_error_at: string | null;
  provider_used: string | null;
}

export interface FailedPagesResponse {
  job_id: number;
  failed_count: number;
  pages: FailedPageInfo[];
}

export interface RetryFailedRequest {
  page_numbers?: number[];
  use_quality_provider?: boolean;
}

export interface RetryFailedResponse {
  job_id: number;
  retrying_pages: number;
  celery_task_id: string | null;
}

export interface ReextractPageResponse {
  job_id: number;
  page_number: number;
  celery_task_id: string;
  provider: string;
}

// ===== Service =====

export const ocrService = {
  /**
   * Upload a PDF file for OCR processing.
   *
   * @param file - PDF file to upload
   * @param targetModuleId - Optional target module ID
   * @param quality - "fast" (Qwen 32B) or "quality" (Qwen 72B)
   */
  uploadPdf: async (
    file: File,
    targetModuleId?: number,
    quality: OCRQuality = 'fast'
  ): Promise<OCRJob> => {
    const formData = new FormData();
    formData.append('file', file);
    if (targetModuleId) {
      formData.append('target_module_id', targetModuleId.toString());
    }
    // Always use OpenRouter as the primary provider
    formData.append('provider', 'openrouter');
    formData.append('quality', quality);

    const response = await api.post<OCRJob>('/ocr/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  },

  /**
   * List OCR jobs for current user.
   */
  listJobs: async (params?: {
    page?: number;
    page_size?: number;
    status?: OCRJobStatus;
  }): Promise<OCRJobListResponse> => {
    const queryParams = new URLSearchParams();
    if (params?.page) queryParams.set('page', params.page.toString());
    if (params?.page_size) queryParams.set('page_size', params.page_size.toString());
    if (params?.status) queryParams.set('status', params.status);

    const response = await api.get<OCRJobListResponse>(
      `/ocr/jobs?${queryParams.toString()}`
    );
    return response.data;
  },

  /**
   * Get OCR job details.
   */
  getJob: async (jobId: number): Promise<OCRJob> => {
    const response = await api.get<OCRJob>(`/ocr/jobs/${jobId}`);
    return response.data;
  },

  /**
   * Cancel a pending or processing job.
   */
  cancelJob: async (jobId: number): Promise<{ message: string; job_id: number }> => {
    const response = await api.post<{ message: string; job_id: number }>(
      `/ocr/jobs/${jobId}/cancel`
    );
    return response.data;
  },

  /**
   * Resume a stuck processing job (e.g., after server restart).
   * Continues from where it left off, skipping already processed pages.
   */
  resumeJob: async (jobId: number): Promise<{
    message: string;
    job_id: number;
    processed_pages: number;
    total_pages: number;
    celery_task_id: string;
  }> => {
    const response = await api.post<{
      message: string;
      job_id: number;
      processed_pages: number;
      total_pages: number;
      celery_task_id: string;
    }>(`/ocr/jobs/${jobId}/resume`);
    return response.data;
  },

  /**
   * List extracted questions for a job.
   */
  listQuestions: async (
    jobId: number,
    params?: {
      page?: number;
      page_size?: number;
      status?: QuestionReviewStatus;
      needs_answer?: boolean;
      needs_image?: boolean;
    }
  ): Promise<ExtractedQuestionListResponse> => {
    const queryParams = new URLSearchParams();
    if (params?.page) queryParams.set('page', params.page.toString());
    if (params?.page_size) queryParams.set('page_size', params.page_size.toString());
    if (params?.status) queryParams.set('status', params.status);
    if (params?.needs_answer !== undefined) {
      queryParams.set('needs_answer', params.needs_answer.toString());
    }
    if (params?.needs_image !== undefined) {
      queryParams.set('needs_image', params.needs_image.toString());
    }

    const response = await api.get<ExtractedQuestionListResponse>(
      `/ocr/jobs/${jobId}/questions?${queryParams.toString()}`
    );
    return response.data;
  },

  /**
   * Get extracted question details.
   */
  getQuestion: async (questionId: number): Promise<ExtractedQuestion> => {
    const response = await api.get<ExtractedQuestion>(`/ocr/questions/${questionId}`);
    return response.data;
  },

  /**
   * Update an extracted question (review/edit).
   */
  updateQuestion: async (
    questionId: number,
    data: QuestionUpdateData
  ): Promise<ExtractedQuestion> => {
    const response = await api.put<ExtractedQuestion>(
      `/ocr/questions/${questionId}`,
      data
    );
    return response.data;
  },

  /**
   * Bulk approve or reject questions.
   */
  bulkReview: async (
    questionIds: number[],
    action: 'approve' | 'reject'
  ): Promise<{ updated: number; new_status: QuestionReviewStatus }> => {
    const response = await api.post<{ updated: number; new_status: QuestionReviewStatus }>(
      '/ocr/questions/bulk-review',
      { question_ids: questionIds, action }
    );
    return response.data;
  },

  /**
   * Import approved questions to a test module.
   */
  importQuestions: async (
    jobId: number,
    targetModuleId: number,
    questionIds?: number[]
  ): Promise<ImportResult> => {
    const response = await api.post<ImportResult>(`/ocr/jobs/${jobId}/import`, {
      target_module_id: targetModuleId,
      question_ids: questionIds,
    });
    return response.data;
  },

  /**
   * Configure test structure before import (saves config without importing).
   */
  configureTest: async (
    jobId: number,
    config: TestConfig
  ): Promise<TestConfigResponse> => {
    const response = await api.post<TestConfigResponse>(
      `/ocr/jobs/${jobId}/configure-test`,
      config
    );
    return response.data;
  },

  /**
   * Import questions and create a new test with configured modules.
   * This is the recommended flow for creating tests from OCR.
   */
  importWithTest: async (
    jobId: number,
    request: ImportWithTestRequest
  ): Promise<ImportWithTestResponse> => {
    const response = await api.post<ImportWithTestResponse>(
      `/ocr/jobs/${jobId}/import-with-test`,
      request
    );
    return response.data;
  },

  /**
   * Get the URL for a PDF page image (for cropping).
   * Returns authenticated URL for the page image endpoint.
   */
  getPageImageUrl: (jobId: number, pageNumber: number, scale: number = 2.0): string => {
    const baseUrl = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1';
    return `${baseUrl}/ocr/jobs/${jobId}/pages/${pageNumber}/image?scale=${scale}`;
  },

  /**
   * Upload a cropped image for a question.
   */
  uploadQuestionImage: async (
    questionId: number,
    imageBlob: Blob
  ): Promise<{
    question_id: number;
    question_image_url: string;
    needs_image: boolean;
    image_extraction_status: string;
  }> => {
    const formData = new FormData();
    formData.append('file', imageBlob, `crop_${Date.now()}.jpg`);

    const response = await api.post<{
      question_id: number;
      question_image_url: string;
      needs_image: boolean;
      image_extraction_status: string;
    }>(`/ocr/questions/${questionId}/upload-image`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  },

  /**
   * Create WebSocket connection for real-time job progress.
   * Returns a cleanup function to close the connection.
   */
  subscribeToJobProgress: (
    jobId: number,
    onMessage: (event: JobProgressEvent) => void,
    onError?: (error: Event) => void
  ): (() => void) => {
    const baseUrl = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1';
    const wsUrl = baseUrl.replace(/^http/, 'ws') + `/ocr/jobs/${jobId}/ws`;

    const token = localStorage.getItem('token');
    const ws = new WebSocket(wsUrl + (token ? `?token=${token}` : ''));

    ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data) as JobProgressEvent;
        onMessage(data);
      } catch (e) {
        console.error('Failed to parse WebSocket message:', e);
      }
    };

    ws.onerror = (event) => {
      if (onError) {
        onError(event);
      }
    };

    // Return cleanup function
    return () => {
      if (ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING) {
        ws.close();
      }
    };
  },

  /**
   * Get job cost estimate based on quality setting.
   */
  estimateCost: (totalPages: number, quality: OCRQuality = 'fast'): number => {
    // Rough estimates in cents per page (OpenRouter pricing)
    // Fast: Qwen 32B ($0.005 input / $0.022 output per 1K tokens)
    // Quality: Qwen 72B ($0.040 per 1K tokens)
    const costPerPage: Record<OCRQuality, number> = {
      fast: 0.10,     // Qwen 32B - cheapest, good accuracy
      quality: 0.25,  // Qwen 72B - better accuracy for complex math
    };
    return Math.round(totalPages * costPerPage[quality]);
  },

  /**
   * Format domain for display.
   */
  formatDomain: (domain: QuestionDomain | null): string => {
    if (!domain) return 'Unknown';
    const mapping: Record<QuestionDomain, string> = {
      algebra: 'Algebra',
      advanced_math: 'Advanced Math',
      geometry_trigonometry: 'Geometry & Trigonometry',
      problem_solving_data_analysis: 'Problem Solving & Data Analysis',
      craft_and_structure: 'Craft & Structure',
      information_and_ideas: 'Information & Ideas',
      expression_of_ideas: 'Expression of Ideas',
      standard_english_conventions: 'Standard English Conventions',
    };
    return mapping[domain] || domain;
  },

  /**
   * Format status for display.
   */
  formatStatus: (status: OCRJobStatus): string => {
    const mapping: Record<OCRJobStatus, string> = {
      pending: 'Pending',
      uploading: 'Uploading',
      processing: 'Processing OCR',
      structuring: 'Structuring',
      review: 'Ready for Review',
      importing: 'Importing',
      completed: 'Completed',
      failed: 'Failed',
      cancelled: 'Cancelled',
    };
    return mapping[status] || status;
  },

  /**
   * Get status color for badges.
   */
  getStatusColor: (
    status: OCRJobStatus
  ): 'default' | 'secondary' | 'destructive' | 'outline' => {
    switch (status) {
      case 'completed':
        return 'default';
      case 'review':
        return 'secondary';
      case 'failed':
      case 'cancelled':
        return 'destructive';
      default:
        return 'outline';
    }
  },

  // ===== Passage Methods =====

  /**
   * List extracted passages for a job.
   */
  listPassages: async (
    jobId: number,
    params?: {
      page?: number;
      page_size?: number;
      status?: QuestionReviewStatus;
    }
  ): Promise<ExtractedPassageListResponse> => {
    const queryParams = new URLSearchParams();
    if (params?.page) queryParams.set('page', params.page.toString());
    if (params?.page_size) queryParams.set('page_size', params.page_size.toString());
    if (params?.status) queryParams.set('status', params.status);

    const response = await api.get<ExtractedPassageListResponse>(
      `/ocr/jobs/${jobId}/passages?${queryParams.toString()}`
    );
    return response.data;
  },

  /**
   * Get extracted passage details.
   */
  getPassage: async (passageId: number): Promise<ExtractedPassage> => {
    const response = await api.get<ExtractedPassage>(`/ocr/passages/${passageId}`);
    return response.data;
  },

  /**
   * Update an extracted passage (review/edit).
   */
  updatePassage: async (
    passageId: number,
    data: PassageUpdateData
  ): Promise<ExtractedPassage> => {
    const response = await api.put<ExtractedPassage>(
      `/ocr/passages/${passageId}`,
      data
    );
    return response.data;
  },

  // ===== Phase 6: Error Handling Methods =====

  /**
   * List failed pages for a job.
   */
  listFailedPages: async (jobId: number): Promise<FailedPagesResponse> => {
    const response = await api.get<FailedPagesResponse>(
      `/ocr/jobs/${jobId}/failed-pages`
    );
    return response.data;
  },

  /**
   * Retry processing for failed pages.
   */
  retryFailedPages: async (
    jobId: number,
    request?: RetryFailedRequest
  ): Promise<RetryFailedResponse> => {
    const response = await api.post<RetryFailedResponse>(
      `/ocr/jobs/${jobId}/retry-failed`,
      request || {}
    );
    return response.data;
  },

  /**
   * Re-extract a specific page with the quality provider.
   * Use when extraction quality is poor and you want to retry with a better model.
   */
  reextractPage: async (
    jobId: number,
    pageNumber: number,
    useQualityProvider: boolean = true
  ): Promise<ReextractPageResponse> => {
    const response = await api.post<ReextractPageResponse>(
      `/ocr/jobs/${jobId}/reextract-page`,
      {
        page_number: pageNumber,
        use_quality_provider: useQualityProvider,
      }
    );
    return response.data;
  },

  /**
   * List pages that were skipped (not detected as question pages).
   */
  listSkippedPages: async (jobId: number): Promise<{
    job_id: number;
    skipped_count: number;
    pages: Array<{
      page_number: number;
      text_preview: string | null;
      text_length: number;
    }>;
  }> => {
    const response = await api.get(`/ocr/jobs/${jobId}/skipped-pages`);
    return response.data;
  },

  /**
   * Force re-process skipped pages to extract questions.
   * This runs the structuring step on pages that already have OCR text.
   */
  processSkippedPages: async (
    jobId: number,
    pageNumbers?: number[]
  ): Promise<{
    message: string;
    job_id: number;
    page_numbers: number[];
    celery_task_id: string;
  }> => {
    const response = await api.post(`/ocr/jobs/${jobId}/process-skipped`, {
      page_numbers: pageNumbers,
    });
    return response.data;
  },
};

export default ocrService;

File: backend/app/models/ocr.py
"""OCR Processing Models for PDF-to-Test Import Pipeline."""

from datetime import UTC, datetime
from typing import TYPE_CHECKING

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    JSON,
    LargeBinary,
    String,
    Text,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin
from app.models.enums import (
    OCRJobStatus,
    OCRProvider,
    QuestionReviewStatus,
    QuestionDomain,
    QuestionDifficulty,
    QuestionType,
)

if TYPE_CHECKING:
    from app.models.user import User
    from app.models.test import TestModule, Passage


class OCRJob(Base, TimestampMixin):
    """
    Represents a PDF processing job.

    Tracks the overall progress of converting a PDF into structured questions.
    Supports checkpoint/resume through page-level tracking.
    """

    __tablename__ = "ocr_jobs"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)

    # Who initiated the job
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    # Target module for import (optional until import step)
    target_module_id: Mapped[int | None] = mapped_column(
        ForeignKey("test_modules.id", ondelete="SET NULL"), index=True
    )

    # Job status
    status: Mapped[OCRJobStatus] = mapped_column(
        Enum(OCRJobStatus, values_callable=lambda x: [e.value for e in x]),
        default=OCRJobStatus.PENDING, nullable=False
    )

    # PDF file info
    pdf_filename: Mapped[str] = mapped_column(String(255), nullable=False)
    pdf_s3_key: Mapped[str] = mapped_column(String(500), nullable=False)
    pdf_hash: Mapped[str] = mapped_column(String(64), index=True)  # MD5 for dedup
    total_pages: Mapped[int] = mapped_column(Integer, nullable=False)

    # Processing progress
    processed_pages: Mapped[int] = mapped_column(Integer, default=0)
    question_pages: Mapped[int] = mapped_column(Integer, default=0)  # Pages with questions
    skipped_pages: Mapped[int] = mapped_column(Integer, default=0)  # Non-question pages

    # Extracted counts
    extracted_questions: Mapped[int] = mapped_column(Integer, default=0)
    approved_questions: Mapped[int] = mapped_column(Integer, default=0)
    imported_questions: Mapped[int] = mapped_column(Integer, default=0)

    # Provider configuration
    ocr_provider: Mapped[OCRProvider] = mapped_column(
        Enum(OCRProvider, values_callable=lambda x: [e.value for e in x]),
        default=OCRProvider.HYBRID
    )

    # Cost tracking (in USD cents)
    estimated_cost_cents: Mapped[int] = mapped_column(Integer, default=0)
    actual_cost_cents: Mapped[int] = mapped_column(Integer, default=0)

    # Timing
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    # Error tracking
    error_message: Mapped[str | None] = mapped_column(Text)
    last_error_page: Mapped[int | None] = mapped_column(Integer)
    retry_count: Mapped[int] = mapped_column(Integer, default=0)

    # Celery task tracking
    celery_task_id: Mapped[str | None] = mapped_column(String(255), index=True)

    # Test configuration for import
    # Format: [{"test_title": "...", "test_type": "full_test|section_test|module_test",
    #           "section": "reading_writing|math|null", "modules": [...]}]
    test_configs: Mapped[list[dict] | None] = mapped_column(JSON)

    # IDs of tests created from this job
    created_test_ids: Mapped[list[int] | None] = mapped_column(JSON)

    # Phase 6: Processing mode and error tracking
    # Values: 'speed' (gpt-4o-mini only), 'quality' (olmOCR only), 'balanced' (fallback)
    processing_mode: Mapped[str] = mapped_column(String(20), default="balanced")
    failed_pages_count: Mapped[int] = mapped_column(Integer, default=0)

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="ocr_jobs")
    target_module: Mapped["TestModule | None"] = relationship("TestModule")
    pages: Mapped[list["OCRJobPage"]] = relationship(
        "OCRJobPage", back_populates="job", cascade="all, delete-orphan",
        order_by="OCRJobPage.page_number"
    )
    questions: Mapped[list["ExtractedQuestion"]] = relationship(
        "ExtractedQuestion", back_populates="job", cascade="all, delete-orphan"
    )
    passages: Mapped[list["ExtractedPassage"]] = relationship(
        "ExtractedPassage", back_populates="job", cascade="all, delete-orphan"
    )

    @property
    def progress_percent(self) -> float:
        """Calculate processing progress as percentage."""
        if self.total_pages == 0:
            return 0.0
        return round((self.processed_pages / self.total_pages) * 100, 1)

    @property
    def duration_seconds(self) -> int | None:
        """Calculate processing duration."""
        if not self.started_at:
            return None
        end = self.completed_at or datetime.now(UTC)
        return int((end - self.started_at).total_seconds())


class OCRJobPage(Base, TimestampMixin):
    """
    Per-page OCR results.

    Stores intermediate results for checkpoint/resume functionality.
    Allows reprocessing individual pages without redoing the entire PDF.
    """

    __tablename__ = "ocr_job_pages"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    job_id: Mapped[int] = mapped_column(
        ForeignKey("ocr_jobs.id", ondelete="CASCADE"), nullable=False, index=True
    )

    # Page info
    page_number: Mapped[int] = mapped_column(Integer, nullable=False)

    # OCR output
    ocr_markdown: Mapped[str | None] = mapped_column(Text)

    # Classification
    is_question_page: Mapped[bool] = mapped_column(Boolean, default=False)

    # Detected figures (graphs, charts, diagrams)
    # Format: [{"label": "parabola graph", "bbox": [y_min, x_min, y_max, x_max], "s3_key": "..."}]
    detected_figures: Mapped[list[dict] | None] = mapped_column(JSON)

    # Page image S3 key (high-res for figure cropping)
    page_image_s3_key: Mapped[str | None] = mapped_column(String(500))

    # Page image data (JPEG bytes stored directly in database)
    page_image_data: Mapped[bytes | None] = mapped_column(LargeBinary)

    # Processing status
    ocr_completed: Mapped[bool] = mapped_column(Boolean, default=False)
    structuring_completed: Mapped[bool] = mapped_column(Boolean, default=False)

    # Cost tracking for this page
    ocr_cost_cents: Mapped[int] = mapped_column(Integer, default=0)
    structuring_cost_cents: Mapped[int] = mapped_column(Integer, default=0)

    # Error handling
    error_message: Mapped[str | None] = mapped_column(Text)
    retry_count: Mapped[int] = mapped_column(Integer, default=0)

    # Phase 6: Enhanced error tracking
    last_error_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    provider_used: Mapped[str | None] = mapped_column(String(50))

    # Relationships
    job: Mapped["OCRJob"] = relationship("OCRJob", back_populates="pages")
    questions: Mapped[list["ExtractedQuestion"]] = relationship(
        "ExtractedQuestion", back_populates="source_page", cascade="all, delete-orphan"
    )
    passages: Mapped[list["ExtractedPassage"]] = relationship(
        "ExtractedPassage", back_populates="source_page"
    )


class ExtractedQuestion(Base, TimestampMixin):
    """
    Extracted question pending review and import.

    Mirrors the Question model structure but with additional fields
    for review workflow and provenance tracking.
    """

    __tablename__ = "extracted_questions"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    job_id: Mapped[int] = mapped_column(
        ForeignKey("ocr_jobs.id", ondelete="CASCADE"), nullable=False, index=True
    )
    source_page_id: Mapped[int] = mapped_column(
        ForeignKey("ocr_job_pages.id", ondelete="CASCADE"), nullable=False, index=True
    )

    # Review status
    review_status: Mapped[QuestionReviewStatus] = mapped_column(
        Enum(QuestionReviewStatus, values_callable=lambda x: [e.value for e in x]),
        default=QuestionReviewStatus.PENDING
    )
    reviewed_by_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL")
    )
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    # Confidence scoring (0.0 - 1.0)
    extraction_confidence: Mapped[float] = mapped_column(Float, default=0.0)
    answer_confidence: Mapped[float] = mapped_column(Float, default=0.0)

    # ===== Question Fields (matches Question model) =====

    # Question content
    question_text: Mapped[str] = mapped_column(Text, nullable=False)
    question_type: Mapped[QuestionType] = mapped_column(
        Enum(QuestionType),  # Uses UPPERCASE names to match existing DB enum
        default=QuestionType.MULTIPLE_CHOICE
    )

    # Image support
    question_image_s3_key: Mapped[str | None] = mapped_column(String(500))
    question_image_url: Mapped[str | None] = mapped_column(String(500))

    # Passage for EBRW questions
    passage_text: Mapped[str | None] = mapped_column(Text)

    # Chart/table data (extracted from OCR)
    chart_title: Mapped[str | None] = mapped_column(String(255))
    chart_data: Mapped[str | None] = mapped_column(Text)  # HTML table from OCR (legacy)

    # Structured table data (new format)
    # Format: {"headers": ["Col1", "Col2"], "rows": [["val1", "val2"]], "title": "..."}
    table_data: Mapped[dict | None] = mapped_column(JSON)

    # Options: [{"id": "A", "text": "...", "image_url": null}, ...]
    options: Mapped[list[dict] | None] = mapped_column(JSON)

    # Correct answer(s)
    correct_answer: Mapped[list[str] | None] = mapped_column(JSON)
    needs_answer: Mapped[bool] = mapped_column(Boolean, default=False)  # Model couldn't determine

    # Explanation
    explanation: Mapped[str | None] = mapped_column(Text)

    # Metadata (UPPERCASE names to match existing DB enums)
    difficulty: Mapped[QuestionDifficulty | None] = mapped_column(Enum(QuestionDifficulty))
    domain: Mapped[QuestionDomain | None] = mapped_column(Enum(QuestionDomain))
    skill_tags: Mapped[list[str] | None] = mapped_column(JSON)

    # Flag for images that need manual review
    needs_image: Mapped[bool] = mapped_column(Boolean, default=False)
    image_extraction_status: Mapped[str | None] = mapped_column(String(50))  # success, failed, manual

    # Validation errors from schema check
    validation_errors: Mapped[list[str] | None] = mapped_column(JSON)

    # Reference to imported question (after successful import)
    imported_question_id: Mapped[int | None] = mapped_column(
        ForeignKey("questions.id", ondelete="SET NULL"), index=True
    )

    # Link to extracted passage (for EBRW questions)
    extracted_passage_id: Mapped[int | None] = mapped_column(
        ForeignKey("extracted_passages.id", ondelete="SET NULL"), index=True
    )

    # Relationships
    job: Mapped["OCRJob"] = relationship("OCRJob", back_populates="questions")
    source_page: Mapped["OCRJobPage"] = relationship("OCRJobPage", back_populates="questions")
    reviewed_by: Mapped["User | None"] = relationship("User", foreign_keys=[reviewed_by_id])
    extracted_passage: Mapped["ExtractedPassage | None"] = relationship(
        "ExtractedPassage", back_populates="questions"
    )

    @property
    def overall_confidence(self) -> float:
        """Combined confidence score."""
        if self.needs_answer:
            return self.extraction_confidence * 0.5  # Penalize missing answers
        return (self.extraction_confidence + self.answer_confidence) / 2

    @property
    def is_ready_for_import(self) -> bool:
        """Check if question is ready for import."""
        return (
            self.review_status == QuestionReviewStatus.APPROVED
            and self.correct_answer is not None
            and not self.needs_answer
            and not self.validation_errors
        )


class ExtractedPassage(Base, TimestampMixin):
    """
    Extracted passage pending review and import.

    Passages are extracted separately from questions for EBRW sections.
    Multiple questions can reference the same passage.
    """

    __tablename__ = "extracted_passages"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    job_id: Mapped[int] = mapped_column(
        ForeignKey("ocr_jobs.id", ondelete="CASCADE"), nullable=False, index=True
    )
    source_page_id: Mapped[int | None] = mapped_column(
        ForeignKey("ocr_job_pages.id", ondelete="SET NULL"), index=True
    )

    # Content
    title: Mapped[str | None] = mapped_column(String(255))
    content: Mapped[str] = mapped_column(Text, nullable=False)
    source: Mapped[str | None] = mapped_column(String(255))  # Publication source
    author: Mapped[str | None] = mapped_column(String(255))
    word_count: Mapped[int | None] = mapped_column(Integer)

    # Figures/Images associated with passage
    # Format: [{"s3_key": "...", "url": "...", "alt": "...", "caption": "..."}]
    figures: Mapped[list[dict] | None] = mapped_column(JSON)

    # Classification
    genre: Mapped[str | None] = mapped_column(String(100))
    topic_tags: Mapped[list[str] | None] = mapped_column(JSON)

    # Review status
    review_status: Mapped[QuestionReviewStatus] = mapped_column(
        Enum(QuestionReviewStatus, values_callable=lambda x: [e.value for e in x]),
        default=QuestionReviewStatus.PENDING
    )
    extraction_confidence: Mapped[float] = mapped_column(Float, default=0.0)

    # Reference to imported passage (after successful import)
    imported_passage_id: Mapped[int | None] = mapped_column(
        ForeignKey("passages.id", ondelete="SET NULL"), index=True
    )

    # Relationships
    job: Mapped["OCRJob"] = relationship("OCRJob", back_populates="passages")
    source_page: Mapped["OCRJobPage | None"] = relationship(
        "OCRJobPage", back_populates="passages"
    )
    questions: Mapped[list["ExtractedQuestion"]] = relationship(
        "ExtractedQuestion", back_populates="extracted_passage"
    )
    imported_passage: Mapped["Passage | None"] = relationship("Passage")

    @property
    def is_ready_for_import(self) -> bool:
        """Check if passage is ready for import."""
        return (
            self.review_status == QuestionReviewStatus.APPROVED
            and self.content
            and len(self.content.strip()) > 0
        )
